src/net_processing.cpp

```diff
@@ -2777,11 +2787,28 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
         std::vector<CAddress> vAddrOk;
         int64_t nNow = GetAdjustedTime();
         int64_t nSince = nNow - 10 * 60;
+
+        // Update/increment addr rate limiting bucket.
+        const auto current_time = GetTime<std::chrono::microseconds>();
+        if (peer->m_addr_token_bucket < MAX_ADDR_TO_SEND) {
+            // Don't increment bucket if it's already full
+            auto time_diff = std::max(current_time - peer->m_addr_token_timestamp, 0us);
+            double increment = std::chrono::duration<double>(time_diff).count() * MAX_ADDR_RATE_PER_SECOND;
+            peer->m_addr_token_bucket = std::min<double>(peer->m_addr_token_bucket + increment, MAX_ADDR_TO_SEND);
+        }
+        peer->m_addr_token_timestamp = current_time;
+
+        Shuffle(vAddr.begin(), vAddr.end(), FastRandomContext());
         for (CAddress& addr : vAddr)
         {
             if (interruptMsgProc)
                 return;
 
+            // Apply rate limiting.
+            if (!pfrom.HasPermission(NetPermissionFlags::Addr)) {
+                if (peer->m_addr_token_bucket < 1.0) continue;
```

Comment:
Should that not be `break;`? `peer->m_addr_token_bucket` is not going to increase inside this loop.

Reply:
I don't understand why it should be `break` - am I missing something? There are other things happening in the loop as well

Reply:
Indeed, a break is enough. Fixed.

@glozow The loop does many things, but if if at some point the rate-limiting condition is reached, it will be reached for all further loop iterations too.